rfc-123-github-actions-ci.md

```diff
@@ -50,6 +50,20 @@ This would be a gradual process, because we have many repos, but this decision w
 - Another system to learn, as well as the replatforming work already ongoing.
 - Duplication of build parameters and steps between all our apps, as they’re all Rails apps.
 
+### Findings from some initial explorations into using GitHub Actions
+
+#### `push` vs `pull_request` build triggers
+
+There has been some [chatter](https://github.com/alphagov/govuk-developer-docs/pull/2497#discussion_r415572147) about if we want `push` or `pull_request` triggers for tests to run, or both. The GOV.UK Developer Tools group has done some investigation into the differences between this, as there was some concern about if we wanted long-running builds (in the case of apps with large test suites, like whitehall) to run twice when not needed. There is not currently any way to differentiate and say "run on PRs from forks" and "run on push events to main repo branches". We'd have really liked a `pull_request_only_from_forks` setting, but that doesn't exist. It's worth noting that this isn't a problem _only_ with GitHub Actions: [Travis](https://github.com/travis-ci/travis-ci/issues/1147) has had the same pain points for years, and we currently do double builds for the apps we use Travis for.
+
+Some solutions to this include:
+
+- Running `on: [pull_request, push]`. This means that internal devs get feedback on their commits when they push them, for example if they've forgotten to run RuboCop locally. And also so that external contributors PRs get CI feedback on their changes. It does mean that for long-running test suites, we could waste compute power, time and the planet's precious natural resources running things twice that don't need to be run twice.
+
+- Running `on: push`. Actions don't run [by default on forked PRs](https://github.com/alphagov/govuk-docker/pull/337) - it's a security feature that GitHub has implemented. And [when a PR is raised from a fork](https://github.com/alphagov/govuk-docker/pull/337) and the receiving repo only has a push trigger, the tests don't run. So we'd be back to the annoying state that is [our current process for forked PRs][merge-a-pull-request].
+
+- Running `on: pull_request`. Functionally, we run the same tests across the board, whichever of these we choose. But this one has the advantage that it's consistent behaviour for internal and external contributors. But internally we'd have to put up with not having tests run on branches (apart from `master`, we should probably include the `master` branch all the time), which would mean changing developer workflows to run more tests locally (ie RuboCop, a common thing that gets forgotten about), or being less worried about "exposing to the world" our force-pushes to get things right on the PR. We would also probably benefit from opening PRs for long-running branches earlier, using the draft PR functionality, and this would encourage that. This seems like the best option to _Issy_, but please feel free to disagree.
+
```

Comment:
Another alternative is wrapping the job in a conditional: https://github.community/t5/GitHub-Actions/don-t-run-actions-on-pull-request-from-fork/td-p/45499

```
jobs:
  prcheck:
    if: github.actor == 'repo-owner'
    runs-on: [ubuntu-latest]
    # etc
```